src/cpp/rtps/messages/MessageReceiver.cpp

```diff
@@ -864,46 +864,40 @@ bool MessageReceiver::proc_Submsg_Data(
         }
 
         payload_size = smh->submessageLength - submsg_no_payload_size;
-
-        if (dataFlag)
+        uint32_t next_pos = msg->pos + payload_size;
+        if (msg->length >= next_pos && payload_size > 0)
         {
-            uint32_t next_pos = msg->pos + payload_size;
-            if (msg->length >= next_pos && payload_size > 0)
+            FASTDDS_TODO_BEFORE(3, 1, "Pass keyFlag in serializedPayload, and always pass input data upwards");
+            if (dataFlag)
             {
                 ch.serializedPayload.data = &msg->buffer[msg->pos];
                 ch.serializedPayload.length = payload_size;
                 ch.serializedPayload.max_size = payload_size;
-                msg->pos = next_pos;
             }
             else
```

Comment:
Why remove `else if (keyFlag)`?

Reply:
Because

1. We are inside an `if (dataFlag || keyFlag)` (line 850)
2. They cannot be both true (check done in line 762)

I will add a comment

Reply:
750eaa0